### DISCUSSION TITLE
Energy MQTT Topic Ac/L1/Energy/Reverse not working as expected

### DISCUSSION THREAD

**Discussion by patatman** [freakent/dbus-mqtt-devices#92] [General]
### Version

0.6.4 (current)

### What version of VenusOS are you running on your GX device

3.20 (current release)

### Does this issue occur during installation or at runtime ?

Runtime

### What is the issue you are actually experiencing?

I'm trying to create a small script which reads data from DSMR-reader send to a MQTT broker, and send it to the victron MQTT broker.
So far so good, and I've mapped out most of the values correctly. I've followed I've read: https://github.com/victronenergy/venus/wiki/dbus#grid-and-genset-meter  to map all the data I have from DSMR to the corresponding Victron topics.

After running the script the entire day, I noticed my Victron values weren't updated when I was sending power back to the grid.

The snippet sending data to the topic in Victron is:
```
        elif msg.topic == f"{DSMR_TOPIC}/reading/phase_currently_returned_l1":
            wattage = float(msg.payload.decode("utf-8")) * 1000
            publish_to_victron(f"W/{portal_id}/grid/{device_instance}/Ac/L1/Energy/Reverse", wattage)
```
This value is in Kw, so I multiply it by 1000 to get the W value, which I then publish to the Victron bus on `Ac/L1/Energy/Reverse` topic. But this doesn't seem to do anything.

If I manually try to send the message on this topic, nothing happens.
The only way I can get the victron to display a negative number, is by sending data to the `/Ac/L1/Energy/Forward` or `/Ac/L1/Power` with a negative sign in front of it (e.g. -2).

My main questions:
 - What is the correct way to use the Reverse topic?
 -  Should I calculate the negative number, and publish that to the `/Ac/L1/Energy/Forward` topic?

It might be a stupid mistake, so sorry if this confuses you.

**Comment by freakent** [comment]
I'm not familar with sending grid data to dbus, but usually you have to send data as json payloads. for example 
```
{
    "value": 23.2
}
```

You appear to be sending a float instead of a json string.

Did you customise the services.yml file? If so please upload a copy here. And a code snippet showing how you are regsitering you device to the device/*/Status topic would be appreciated too.

If you are new to this, I would always advise you start with something simple like creating a test Temperature device. There are examples in the samples folder. You can also debug most issues using MQTT Explorer.

(moving this to [Discussions](https://github.com/freakent/dbus-mqtt-devices/discussions) as it doesn't appear to be a driver defect)

**Reply by patatman** [reply]
Sorry for placing this in the wrong category.
I'm sending it as Json, but in a later stage as a separate function:
```
def publish_to_victron(topic, value):
    payload = json.dumps({"value": value})
    victron_client.publish(topic, payload)
    print(f"Published to Victron: {topic} with payload {payload}")
```
For example a message published would be:
`Published to Victron: W/b827eba77875/grid/1/Ac/L1/Energy/Reverse with payload {"value": 0.0}`

I see all values correctly on the bus, and it works for all my loads.
But for some reason as soon as I start delivering power back to the grid, it doesn't turn negative.

I didn't customize the `services.yml`

Below is the snippet for registering:
```
# Callback Functions for Victron MQTT
def on_connect_victron(client, userdata, flags, reason_code, properties):
    print(f"Connected to Victron MQTT with result code {reason_code}")
    client.subscribe(subscribe_topic)
    client.publish(status_topic, connected_payload)

def on_message_victron(client, userdata, msg):
    global portal_id, device_instance
    data = json.loads(msg.payload.decode("utf-8"))
    portal_id = data.get("portalId", "")
    device_instance = data.get("deviceInstance", {}).get("meter1", "")
    print(f"got portal id: {portal_id} and device instance: {device_instance}")

    if portal_id and device_instance:
        # Construct the topics
        custom_name_topic = f"W/{portal_id}/grid/{device_instance}/CustomName"
        role_topic = f"W/{portal_id}/grid/{device_instance}/Role"
        # Construct the payloads
        custom_name_payload = json.dumps({"value": CLIENT_ID})
        role_payload = json.dumps({"value": "grid"})
        # Publish the custom name and role
        client.publish(custom_name_topic, custom_name_payload)
        client.publish(role_topic, role_payload)
        print(f"Published custom name to {custom_name_topic} and role to {role_topic}.")
```

And:
```
# Subscribe and Publish Topics
subscribe_topic = f"device/{CLIENT_ID}/DBus"
status_topic = f"device/{CLIENT_ID}/Status"
# Status Message Payloads
connected_payload = json.dumps({"clientId": CLIENT_ID, "connected": 1, "version": "v0.1", "services": {"meter1": "grid"}})
disconnected_payload = json.dumps({"clientId": CLIENT_ID, "connected": 0, "version": "v0.1", "services": {"meter1": "grid"}})
```

I hacked around a bit, and tried different stuff before actually moving to the grid sensor. But I'm by no means an expert in this. 

Steps I took to make sure I have a basic understanding:
1. Listened on the Victron bus for a while to see what's going on.
2. Try to inject some payload to get an ID using MQTT Explorer (publishing to the device/name/Status topic)
3. Start publishing some data on the needed topic to show data in the Victron
4. After figuring it out by hand, start hacking away in Python
5. Watching result, and be satisfied with the initial script
6. See weird behaviour when returning solar.
7. Hack around how to get value to display negative (sending a payload with a minus (-) sign in front)
8. Question if this is the correct way to send payload, or I'm just making a rookie mistake.

Hopefully this clarifies it a bit, and how I got to this point.
I'm going to strip my code of all the hard coded values and push it to github a.s.a.p. so it's easier to understand what's going on.

**Reply by freakent** [reply]
No problem. Now that github has Discussions I'm trying to keep problem investigations separate from issues. Issues are my work list. 

I have read on the Victron Modifications community that there is a bug publishing -ve values through MQTT. See the change log in this post.  

https://community.victronenergy.com/questions/262520/venus-os-v3304-available-for-testing.html

**Reply by patatman** [reply]
Thanks for the fast reply, I'm currently installing 3.21 which should have a fix.
https://professional.victronenergy.com/news/detail/237/

I'm still confused sine they mention negative numbers specifically:
> Fix MQTT issue that prevented writing negative numbers over the MQTT protocol. This bug affected systems with customer-own software, for example special control loops that run a different ESS algorithm than the standard Victron one. The bug was introduced per v3.20, last week.

I'm just publishing the value from DSMR to the `/Ac/L1/Energy/Reverse` topic, which is a positive number e.g. 2.0 (for 2 kw).
Now that I'm thinking of it, this is probably not related to your driver but more on the Victron internals. 
I did some tiny digging in other source codes that seem to work with your driver, and show negative numbers. Just haven't figured out how they are doing it.

For example: https://github.com/NuKeMtje/ESPHomeDSMRVictron?tab=readme-ov-file

Specifically this line:
```
  - platform: template
    name: "Total injected"
    id: total_injected
    lambda: |-
      return id(energy_produced_t1).state + id(energy_produced_t2).state;
    update_interval: 1s
    device_class: energy
    state_class: total_increasing
    unit_of_measurement: kWh
    on_value:
      then:
        - mqtt.publish_json:
            topic: !lambda |-
              return str_sprintf("W/%s/grid/%s/Ac/Energy/Reverse", id(portal_id).c_str(), id(device_instance).c_str());
            payload: |-
              root["value"] = id(total_injected).state;
``` 
Seems to imply it also sends a positive number. 

For now, I'll wait and see what happens after upgrading to 3.21 and report back when the sun is going to start shining.

**Reply by patatman** [reply]
Just a quick update, after upgrading to 3.21 the driver stopped working with an error:
```
@4000000065d468bd28b6eb94 *** starting dbus-mqtt-devices ***
@4000000065d468bf1474cf0c Traceback (most recent call last):
@4000000065d468bf14769044   File "/data/drivers/dbus-mqtt-devices-0.6.4/dbus_mqtt_devices.py", line 34, in <module>
@4000000065d468bf1476ab9c     from device_manager import MQTTDeviceManager
@4000000065d468bf1476bb3c   File "/data/drivers/dbus-mqtt-devices-0.6.4/device_manager.py", line 17, in <module>
@4000000065d468bf1476cec4     import yaml
@4000000065d468bf1476da7c ModuleNotFoundError: No module named 'yaml'
```

So I'm now running the setup script again and seeing if that fixes the issue.
Should I reinstall everytime I upgrade VenusOS?

**Reply by patatman** [reply]
Another update:
After reinstalling everything works again.

Except the reverse topic still doesn't work:
<img width="798" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/d2ab4adb-e008-48bf-9f51-51f6365b040e">

**Reply by freakent** [reply]
What does the Reverse value look like under the N/b827eba77875/grid/1/Ac/L1/Energy/Reverse topic?

**Reply by patatman** [reply]
I don't have a value on that topic. Should I be publishing data on that topic as well?

<img width="622" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/8402601f-bb1a-4db5-8c19-bd2101cb59e1">

**Reply by freakent** [reply]
No, definitely not! The Victron MQTT dbus api has 3 topic braches. W/* is where you write values, N/* (for notification) is where you read values and R/* is for read requests. You will need to send a keepalive message every 60 seconds to read the N values. It's very important to see what the dbus is seeing using the N/* topics. You can also use dbus-spy on the venusos command line to check the dbus values directly.

**Reply by freakent** [reply]
There is a python script in the samples directory that will send keepalives for you, if you need it.

**Reply by patatman** [reply]
Using `dbus-spy` I'm able to see the values:
<img width="1800" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/f77a9bd3-3538-4d96-8ffa-10f94c686ae1">

If I understand you correctly, this means the Dbus and thus driver is working as expected.
And this is probably related to something in VenusOS and not specifically your driver?

**Reply by patatman** [reply]
Just to be sure I updated to the latest Beta, to see if the problem still exists.
And unfortunately it still exist, even though the Dbus information seems to be correct.
<img width="1800" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/88ae826a-d483-4ddd-a8f8-a0becc093086">

The screenshot above is with VenusOS 3.30~7

**Reply by freakent** [reply]
OK dbus seems to be receiving the value, what do you see in the N/* topics?

> After running the script the entire day, I noticed my Victron values weren't updated when I was sending power back to the grid.

Going back to your original question, where did you notice the Victron values weren't updated? which GUI?

**Reply by patatman** [reply]
I noticed the issue in the main interface when it showed 0, and 0.:
<img width="796" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/fe1c5144-90b3-4e6e-aafa-ef6dd4b14507">

After using the keepalive script in the samples dir, i'm getting:
I've included the "device list -> mqttvictron" subpage to show a bit more information, and also that it displays 0. 
<img width="845" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/f5f4785c-892b-49b8-897e-a83639db9979">

Yesterday I did some tests, and the only way to get a negative number in here is to push a negative number (e.g. -5) to the forward topic.

**Reply by freakent** [reply]
Well it looks like the driver is working as it should. The correct values are getting sent to the dbus and being returned to other MQTT subscribers. There is not a lot else I can help you with.  You could try posting a question on the modification community forum. If you explain you can see the values in dbus spy but they are not shown correctly on the GX console that will help to focus in on the issue. Perhaps someone there can explain of these values are handled and displayed in the GX console.

**Reply by patatman** [reply]
That's a good point, and something I'll definitely do.
Thanks for all your help, and guiding me through the troubleshooting steps. Much appreciated.
You can close this discussion to keep it clean, should I get an answer how to send the data correctly I'll adjust the readme or add an sample for other to use as well.

**Reply by freakent** [reply]
If you get a solution add it to this discussion thread so that others may benefit. I'm happy to keep this thread open.

**Reply by patatman** [reply]
I'm a step further, if I look into the inverter itself it looks like it's publishing negative numbers.
<img width="1800" alt="image" src="https://github.com/freakent/dbus-mqtt-devices/assets/15031001/e9419bb1-b253-4839-8f98-2bc8f53f1b93">

I'll adjust my script and see what happens.

**Comment by freakent** [comment]
> Just a quick update, after upgrading to 3.21 the driver stopped working with an error:
> 
> ```
> @4000000065d468bd28b6eb94 *** starting dbus-mqtt-devices ***
> @4000000065d468bf1474cf0c Traceback (most recent call last):
> @4000000065d468bf14769044   File "/data/drivers/dbus-mqtt-devices-0.6.4/dbus_mqtt_devices.py", line 34, in <module>
> @4000000065d468bf1476ab9c     from device_manager import MQTTDeviceManager
> @4000000065d468bf1476bb3c   File "/data/drivers/dbus-mqtt-devices-0.6.4/device_manager.py", line 17, in <module>
> @4000000065d468bf1476cec4     import yaml
> @4000000065d468bf1476da7c ModuleNotFoundError: No module named 'yaml'
> ```
> 
> So I'm now running the setup script again and seeing if that fixes the issue. Should I reinstall everytime I upgrade VenusOS?

You do not need to do a full reinstall, but you will need to run setup.sh again after a venusos install. That's another reason why I want to get rid of the YAML dependency in a future version.

**Reply by patatman** [reply]
Should I report a bug for this? Or is this expected behaviour?
Also willing to add a section in the readme regarding upgrading, if it's necessary to reinstall.

**Reply by freakent** [reply]
I have raised an issue for this already.

**Comment by freakent** [comment]
Thanks for sendng the pull request. I have merged it  into main but it won't go public until I regenerate the templates in the next release.
